cosmwasm/packages/wasmi-runtime/src/wasm/io.rs

```diff
@@ -65,99 +66,89 @@ pub fn encrypt_output(
+
+        // Output is an object
+        // Encrypt all Wasm messages (keeps Bank, Staking, etc.. as is)
+        WasmOutput::OkObject { ok } => {
+            for msg in &mut ok.messages {
+                if let CosmosMsg::Wasm(wasm_msg) = msg {
+                    encrypt_wasm_msg(wasm_msg, nonce, user_public_key)?;
                 }
             }
 
-            if ok["log"].is_array() {
-                if let Value::Array(events) = &mut ok["log"] {
-                    for e in events {
-                        if e["key"].is_string() {
-                            if let Value::String(k) = &mut e["key"] {
-                                e["key"] = encrypt_serializeable(&key, k)?;
-                            }
-                        }
-                        if e["value"].is_string() {
-                            if let Value::String(v) = &mut e["value"] {
-                                e["value"] = encrypt_serializeable(&key, v)?;
-                            }
-                        }
-                    }
-                }
+            // Encrypt all logs
+            for log in &mut ok.log {
+                log.key = encrypt_serializeable(&key, &log.key)?;
+                log.value = encrypt_serializeable(&key, &log.value)?;
             }
 
-            if v["Ok"]["data"].is_string() {
-                if let Value::String(data) = &mut v["Ok"]["data"] {
-                    v["Ok"]["data"] = encrypt_serializeable(&key, data)?;
-                }
+            // If there's data at all
+            if let Some(data) = &mut ok.data {
+                *data = encrypt_serializeable(&key, data)?;
             }
         }
-    }
+    };
+
+    debug!("WasmOutput: {:?}", output);
 
-    let output = serde_json::ser::to_vec(&v).map_err(|err| {
+    // Serialize back to json and return
+    let encrypted_output = serde_json::to_vec(&output).map_err(|err| {
         error!(
             "got an error while trying to serialize output json into bytes {:?}: {}",
-            v, err
+            output, err
         );
         EnclaveError::FailedToSerialize
     })?;
 
-    debug!(
-        "Output after encryption: {:?}",
-        String::from_utf8_lossy(&output)
-    );
+    Ok(encrypted_output)
+}
+
+fn encrypt_wasm_msg(
+    wasm_msg: &mut WasmMsg,
+    nonce: IoNonce,
+    user_public_key: Ed25519PublicKey,
+) -> Result<(), EnclaveError> {
+    match wasm_msg {
+        WasmMsg::Execute { msg, .. } => {
+            let mut msg_to_pass =
+                SecretMessage::from_base64((*msg).to_string(), nonce, user_public_key)?;
+
+            msg_to_pass.encrypt_in_place()?;
+            *msg = b64_encode(&msg_to_pass.to_slice());
+        }
+        WasmMsg::Instantiate { msg, .. } => {
+            let mut msg_to_pass =
+                SecretMessage::from_base64((*msg).to_string(), nonce, user_public_key)?;
+
+            msg_to_pass.encrypt_in_place()?;
+            *msg = b64_encode(&msg_to_pass.to_slice());
+        }
```

Comment:
The section:
```rust
let mut msg_to_pass =
  SecretMessage::from_base64((*msg).to_string(), nonce, user_public_key)?;

msg_to_pass.encrypt_in_place()?;
*msg = b64_encode(&msg_to_pass.to_slice());
```
repeats twice here. you should maybe wrap it up in a function, or (and i don't know if it works) try to write the match arm as:
```
WasmMsg::Execute { msg, .. } | WasmMsg::Instantiate { msg, .. } => { ... }
```

Reply:
```
WasmMsg::Execute { msg, .. } | WasmMsg::Instantiate { msg, .. } => { ... }
```
This has just changed my life, thanks :stuck_out_tongue:

Reply:
awesome, haha :D